packages/gatsby/src/schema/resolvers.ts

```diff
@@ -221,12 +248,56 @@ export function link<TSource, TArgs>(
     TArgs
   >
 ): GatsbyResolver<TSource, TArgs> {
-  return async function linkResolver(
+  return function linkResolver(
     source,
     args,
     context,
     info
-  ): Promise<any> {
+  ): IGatsbyNode | Array<IGatsbyNode> | null | Promise<any> {
+    if (
+      options.by === `id` &&
+      (fieldConfig.resolve || context.defaultFieldResolver) ===
+        defaultFieldResolver
+    ) {
+      return linkResolverDefaultId(source, args, context, info)
+    }
+
+    return linkResolverAsync(source, args, context, info)
+  }
+
+  function linkResolverDefaultId(
+    source,
+    args,
+    context,
+    info
+  ): IGatsbyNode | Array<IGatsbyNode> | null {
+    const fieldValue = defaultFieldResolver(source, args, context, {
+      ...info,
+      from: options.from || info.from,
+      fromNode: options.from ? options.fromNode : info.fromNode,
+    })
+
+    if (fieldValue == null) {
+      return null
+    }
+
+    const returnType = getNullableType(options.type || info.returnType)
+    const type = getNamedType(returnType)
+
+    if (Array.isArray(fieldValue)) {
+      return context.nodeModel.getNodesByIds(
+        { ids: fieldValue, type: type },
+        { path: context.path }
+      )
+    } else {
+      return context.nodeModel.getNodeById(
+        { id: fieldValue, type: type },
+        { path: context.path }
+      )
+    }
```

Comment:
Looks like we duplicate quite a bit of code now. Maybe just test for thenable ourselves and get rid of `async`:

```js
function isThenable(value: any): value is Promise<any> {
  return Boolean(
    value && typeof value === 'object' && typeof value.then === 'function'
  );
}

function linkResolver(/**/) {
  const resolver = fieldConfig.resolve || context.defaultFieldResolver
  const fieldValueOrThenable = resolver(source, args, context, {
    ...info,
    from: options.from || info.from,
    fromNode: options.from ? options.fromNode : info.fromNode,
  })

  if (isThenable(fieldValueOrThenable)) {
    return fieldValueOrThenable.then(fieldValue => linkResolverImpl(fieldValue, source, args, context, info))
  } else {
    return linkResolverImpl(fieldValue, source, args, context, info)
  }
}
```
We'll have to remove the the `async` from `linkResolverImpl` as well and just stick with `then` for `runQuery`.

Reply:
You're right. We can do that.

I did this initially but the result felt very splintered and in the end the duplication looks worse than it actually is. But I'm fine doing it anyways.

Reply:
Refactored the whole thing. Basically jump to the next function sync or async depending on the `?.then`.